### PR TITLE
Add resistor-color exercise

### PR DESCRIPTION
**Pull request by still-flow**
Ported over from the C track, but with a Rust-ic flavor. Hopefully the external crate usage is okay.

### CODE REVIEW

File: exercises/practice/resistor-color/.meta/example.rs
Comment (coriolinus) [inline exercises/practice/resistor-color/.meta/example.rs:19]:
```suggestion
    color.int_value()
```

This improves things in case you ever change the internal representation.

File: exercises/practice/resistor-color/.meta/example.rs
Comment (coriolinus) [inline exercises/practice/resistor-color/.meta/example.rs:25]:
```suggestion
    ResistorColor::from_int(value).map(|color| format!("{:?}", color)).unwrap_or("value out of range".into())
```

Simplify.

File: exercises/practice/resistor-color/.meta/example.rs
Comment (coriolinus) [inline exercises/practice/resistor-color/.meta/example.rs:29]: This seems like a place for the [`enum_iterator` crate](https://docs.rs/enum-iterator/0.7.0/enum_iterator/), if we're using external crates anyway.

However, I have to ask: what learning objective is this test supporting?

File: exercises/practice/resistor-color/src/lib.rs
Comment (coriolinus) [inline exercises/practice/resistor-color/src/lib.rs:12]: Why is this sorted backwards? It should be sorted either forwards, or alphabetically.

**Review by coriolinus — COMMENTED** [review]
Just took a quick look; it's definitely a good start! Some notes:

- You're going to need to edit the top-level `config.json` to add the exercise there as well.
- What is this exercise meant to teach?
- If we can come up with a good educational goal, then could we structure this exercise as a concept exercise instead of a practice?

File: exercises/practice/resistor-color/.meta/example.rs
Comment (still-flow) [reply exercises/practice/resistor-color/.meta/example.rs:27]: I'm not too sure. The C track's version (also Python) all have a test to this end. I guess it's to verify that the enum is represented correctly?

File: exercises/practice/resistor-color/src/lib.rs
Comment (still-flow) [reply exercises/practice/resistor-color/src/lib.rs:12]: I think alphabetically makes sense, so that the student has to read the instructions in order to make correct enum.

File: exercises/practice/resistor-color/.docs/hints.md
Comment (still-flow) [reply exercises/practice/resistor-color/.docs/hints.md:1]: I'll fix the link in the instructions.

**Comment by still-flow** [comment]
Re what this exercise teaches: I think it teaches about C-like enums, also about external crate usage. And the neat "{:?}" formatting trick.

As for making this a concept exercise: sure! It's just that I'm new to contributing exercises, so a practice one is certainly easier. However, I'll probably be able to figure things out if necessary.

**Comment by still-flow** [comment]
Re editing top-level `config.json`: where should I read to learn how to properly do it?

**Comment by still-flow** [comment]
Well, here's my attempt, anyway. Please feel free to correct me.

File: config.json
Comment (coriolinus) [inline config.json:1417]: I see you've placed this last among the non-deprecated exercises. It's not a bad guess, but it should actually go somewhere among the difficulty-4 problems. For practice exercises, the ordering of the exercises in this list determines the "next recommended exercise" flow, so it's best to keep the ordering sensible.

File: config.json
Comment (coriolinus) [inline config.json:1413]: As `enums` is a concept (as defined in the `concepts` section of this file), this should go in the `practices` array instead of `topics`. We have [lots of cleanup](https://github.com/exercism/rust/issues/1407) to do in this repo for that, so let's avoid exacerbating the problem.

File: config.json
Comment (coriolinus) [inline config.json:1411]: I'm curious: why did you assign this as difficulty 4 when the equivalent C exercise has it as difficulty 1?

File: exercises/practice/resistor-color/.docs/instructions.md
Comment (coriolinus) [inline exercises/practice/resistor-color/.docs/instructions.md:7]: Other bulleted lists in this doc use `-` instead of `*` for the bullets.

File: exercises/practice/resistor-color/Cargo.toml
Comment (coriolinus) [inline exercises/practice/resistor-color/Cargo.toml:4]: In general, an exercise which is live on the site should be version `1.0.0` or higher.

File: exercises/practice/resistor-color/.meta/example.rs
Comment (coriolinus) [inline exercises/practice/resistor-color/.meta/example.rs:30]:
```suggestion
        .map(|i| ResistorColor::from_int(i).unwrap())
```

We know from within this file that the range `0..=9` are all resistor colors, and there are no colors outside that range, so there's no need to consider the case where that isn't true. Might be worth adding a comment to that effect as well, justifying `unwrap`.

File: exercises/practice/resistor-color/tests/resistor-color.rs
Comment (coriolinus) [inline exercises/practice/resistor-color/tests/resistor-color.rs:1]:
```suggestion
use resistor_color::{colors, color_to_value, value_to_color_string, ResistorColor};
```

Just makes usage a bit cleaner to import the functions as well.

**Review by coriolinus — COMMENTED** [review]
Some more notes below.

The biggest comment I have is that we have to decide now if this is going to be a concept exercise; it's not something that can be changed after the exercise goes live. This seems plausible to me for teaching the `external_crates` concept. Upside: it gets a lot more visibility, as the concepts in Rust still need a lot of fleshing out. Downside: creating the concept requires a fair amount of extra work. If you do decide to convert it to a concept exercise, it should use the [`enum-iterator` crate](https://crates.io/crates/enum-iterator). 

[Concept docs](https://exercism.org/docs/building/tracks/concepts), [concept exercise docs](https://exercism.org/docs/building/tracks/concept-exercises).

If you don't want to do the extra work, this is also fine as a practice exercise. Just let me know whether you want to convert this or not.

File: config.json
Comment (still-flow) [reply config.json:1411]: I think it's *a bit* more difficult because of the extra work expected. Are there difficulty levels between 1 and 4?

**Comment by still-flow** [comment]
Okay, here's all the fixups for now. I think I should try converting this to a concept exercise, but failing that, this should be good for now.

**Comment by coriolinus** [comment]
Cool, I'll look forward to seeing how this looks as a concept!

**Comment by still-flow** [comment]
Alright, here is a rough first version of that.

File: config.json
Comment (still-flow) [reply config.json]: Changed it to 1 now.

File: concepts/external-crates/about.md
Comment (coriolinus) [inline concepts/external-crates/about.md:3]:
```suggestion
A crate is a compilation unit in Rust. A crate can be compiled into a binary or into a library. Most of the time, you will be using `cargo` to manage external dependencies. Adding a dependency can be as simple as adding a line to your `Cargo.toml` file.
```

I think splitting that sentence into two simplifies things a bit.

File: concepts/external-crates/introduction.md
Comment (coriolinus) [inline concepts/external-crates/introduction.md:9]:
```suggestion
```

Concept docs should stand alone, and there isn't a 1:1 mapping between concepts and concept exercises, so we shouldn't say this.

File: exercises/concept/resistor-color/.meta/exemplar.rs
Comment (coriolinus) [inline exercises/concept/resistor-color/.meta/exemplar.rs:26]: Why not retain the `IntEnum`-based implementation? I'd expect that to be able to produce a faster lookup than iterating through the entire list. I don't think it's a problem for the concept introducing external dependencies to have two different external dependencies.

File: exercises/concept/resistor-color/.docs/introduction.md
Comment (coriolinus) [inline exercises/concept/resistor-color/.docs/introduction.md:9]: It's worth specifying the names of the included external dependencies, and linking to the relevant documentation. IIRC students using the web interface can't ready `Cargo.toml` by default.

File: exercises/concept/resistor-color/src/lib.rs
Comment (coriolinus) [inline exercises/concept/resistor-color/src/lib.rs:27]:
```suggestion
    unimplemented!("return a list of all the colors ordered by resistance")
```

Avoid confusion about i.e. alphabetical order.

**Comment by coriolinus** [comment]
This looks very plausible as a concept to me!

File: exercises/concept/resistor-color/.meta/exemplar.rs
Comment (still-flow) [reply exercises/concept/resistor-color/.meta/exemplar.rs:26]: I thought this was what you meant by 

> it should use the enum-iterator crate

To be honest, I'm not sure whether the two crates would interact nicely.

File: exercises/concept/resistor-color/.docs/introduction.md
Comment (still-flow) [reply exercises/concept/resistor-color/.docs/introduction.md:9]: Isn't it enough to specify that in `instructions.md`?

Comment (coriolinus) [reply exercises/concept/resistor-color/.docs/introduction.md:9]: Oh, you're right, sorry.

File: exercises/concept/resistor-color/.meta/exemplar.rs
Comment (still-flow) [reply exercises/concept/resistor-color/.meta/exemplar.rs:26]: Ok, sure, I'll try!

**Comment by still-flow** [comment]
Alright, re-adding the `int-enum` crate seems to have worked.

File: config.json
Comment (coriolinus) [inline config.json:101]: Note that until the status is `active`, students won't see this exercise. 

The last thing that I think is necessary before setting this active is to ensure that the rust test runner has both of these dependencies in its list of supported external crates.

**Review by coriolinus — APPROVED** [review]
Looks good! Thanks for putting in the effort to push this over the finish line.

File: config.json
Comment (still-flow) [reply config.json:101]: > Note that until the status is active, students won't see this exercise.

Oh, right. But we can't set it to `active` yet, right? Because some files (e.g. `design.md`) are currently missing.

> The last thing that I think is necessary before setting this active is to ensure that the rust test runner has both of these dependencies in its list of supported external crates.

Is that something I can do?

Comment (still-flow) [reply config.json:101]: If I understood correctly, [this line](https://github.com/exercism/rust-test-runner/blob/59acc0d59da1e2c2f6778ceca2e1f9c9da4e2e79/local-registry/supported_crates#L46) and [that line](https://github.com/exercism/rust-test-runner/blob/59acc0d59da1e2c2f6778ceca2e1f9c9da4e2e79/local-registry/supported_crates#L75) indeed show that the requisite crates are supported.

Comment (coriolinus) [reply config.json:101]: Oh very good! It's possible in principle that the versions don't line up--that's a [known issue](https://github.com/exercism/rust-test-runner/issues/37)--but I think we could switch this to active without worrying too much.

Want to do that before I merge, or would you prefer to leave it WIP for some more review first?

Comment (still-flow) [reply config.json:101]: I think squashing commits would be a good idea before merging (or is that something possible on your end?). But otherwise I'd leave that up to your discretion: perhaps more review is not a bad idea.

Comment (coriolinus) [reply config.json:101]: Sure, we use the github's squash and merge by default in this repo, no worries!

In my opinion this is ready to set as active now. Worst case we have to add a PR changing it to WIP if students have problems.

Comment (still-flow) [reply config.json:101]: Alright, if you say it's fine, then let's do this!

**Comment by still-flow** [comment]
It works! Albeit not without trouble: it would seem that the crates have to be re-fetched on every test instance, which slows tests down considerably, even leading to timeouts sometimes. That's probably why there was a rule against external crates?

**Comment by coriolinus** [comment]
Hmm, I'm pretty confident that it's not re-fetching the crates each time, because the test runners run offline. That's why there's a list of supported crates in the first place. However, it's worth looking into the timeout rate.

**Pull request merged**
